File: packages/sdk/src/resources/transactions.ts
import type { LinkOptions } from '@/config';
import { BaseResource } from '@/resources/base';
import type {
  ITransactionsResource,
  ListTransactionsParams,
} from '@/resources/interfaces';
import type { TransactionsPage } from '@/types/index';
import { z } from 'zod';

const transactionSchema = z.looseObject({
  id: z.string(),
  source_id: z.string().nullable(),
  amount: z.number(),
  currency: z.string(),
  created_date: z.string(),
  description: z.string(),
  origin: z.enum(['link', 'external_connection']),
  category: z.string().nullable(),
  status: z.string(),
});
const transactionsPageSchema = z.union([
  z.array(transactionSchema).transform((data) => ({ data })),
  z.looseObject({
    data: z.array(transactionSchema),
    has_more: z.boolean().optional(),
  }),
]);

export class TransactionsResource
  extends BaseResource
  implements ITransactionsResource
{
  constructor(options: LinkOptions) {
    super(options, '/transactions');
  }

  private buildUrl(params: ListTransactionsParams): string {
    const url = new URL(this.endpoint);

    if (params.limit !== undefined) {
      url.searchParams.set('limit', String(params.limit));
    }
    if (params.starting_after !== undefined) {
      url.searchParams.set('starting_after', params.starting_after);
    }
    if (params.ending_before !== undefined) {
      url.searchParams.set('ending_before', params.ending_before);
    }
    if (params.start_date !== undefined) {
      url.searchParams.set('date_start', params.start_date);
    }
    if (params.end_date !== undefined) {
      url.searchParams.set('date_end', params.end_date);
    }
    if (params.category !== undefined) {
      url.searchParams.set('category', params.category);
    }
    if (params.origin !== undefined) {
      url.searchParams.set('origin', params.origin);
    }
    if (params.sources !== undefined) {
      for (const source of params.sources) {
        url.searchParams.append('sources[]', source);
      }
    }

    return url.toString();
  }

  async list(params: ListTransactionsParams = {}): Promise<TransactionsPage> {
    const { status, data, rawBody } = await this.apiFetch({
      method: 'GET',
      url: this.buildUrl(params),
    });

    if (status < 200 || status >= 300) {
      this.throwApiError('list transactions', status, data, rawBody);
    }

    return this.parseResponse(
      'list transactions',
      status,
      () => transactionsPageSchema.parse(data) as TransactionsPage,
    );
  }
}

File: packages/sdk-go/transactions.go
package link

import (
	"bytes"
	"context"
	"encoding/json"
	"net/http"
	"net/url"
	"strconv"
)

// TransactionsResource provides transaction operations.
type TransactionsResource struct {
	base *baseResource
}

// List returns a page of transactions.
func (r *TransactionsResource) List(ctx context.Context, params *ListTransactionsParams) (*TransactionsPage, error) {
	endpoint, err := url.Parse(r.base.baseURL + "/transactions")
	if err != nil {
		return nil, newConfigurationError("Invalid transactions URL: " + err.Error())
	}
	query := endpoint.Query()
	if params != nil {
		if params.Limit != nil {
			query.Set("limit", strconv.FormatInt(*params.Limit, 10))
		}
		if params.StartingAfter != nil {
			query.Set("starting_after", *params.StartingAfter)
		}
		if params.EndingBefore != nil {
			query.Set("ending_before", *params.EndingBefore)
		}
		if params.StartDate != nil {
			query.Set("date_start", *params.StartDate)
		}
		if params.EndDate != nil {
			query.Set("date_end", *params.EndDate)
		}
		if params.Category != nil {
			query.Set("category", *params.Category)
		}
		if params.Origin != nil {
			query.Set("origin", string(*params.Origin))
		}
		for _, source := range params.Sources {
			query.Add("sources[]", source)
		}
	}
	endpoint.RawQuery = query.Encode()
	response, err := r.base.fetch(ctx, http.MethodGet, endpoint.String(), nil, nil)
	if err != nil {
		return nil, err
	}
	if response.status < 200 || response.status >= 300 {
		return nil, newAPIError("list transactions", response.status, response.data, response.rawBody)
	}
	data := bytes.TrimSpace([]byte(response.rawBody))
	var page TransactionsPage
	if len(data) > 0 && data[0] == '[' {
		if err := json.Unmarshal(data, &page.Data); err != nil {
			return nil, newResponseError("list transactions", response.status, err)
		}
	} else {
		if err := decodeResponse("list transactions", response, &page); err != nil {
			return nil, err
		}
	}
	return &page, nil
}
